internal/zinc-apiinfo/src/test/scala/sbt/internal/inc/ClassCanonicalNameSpec.scala

```diff
@@ -0,0 +1,96 @@
+      run(refClass) === full &&
+      run(tagClass) === full &&
+      run2(refClass) === canon &&
+      run2(tagClass) === canon
+    )
+  }
+
+  // O means object, C means cleans, therefore..
+  // OOO means object { object { object } }
+  // CCC means class { class { class } }
+  // etc
+  object O extends Name("x$")
+  object C extends Name("x")
+
+  object OO extends Name("x.y$")
+  object OC extends Name("x.y")
+  object CO extends Name("x.y$")
+  object CC extends Name("x.y")
+
+  object OOO extends Name("x.y$$z$", malformedCN = true)
+  object OOC extends Name("x.y$$z", malformedCN = true)
+  object OCO extends Name("x.y.z$")
+  object OCC extends Name("x.y.z")
+  object COO extends Name("x.y$$z$", malformedCN = true)
+  object COC extends Name("x.y$$z", malformedCN = true)
+  object CCO extends Name("x.y.z$")
+  object CCC extends Name("x.y.z")
+
+  def data = Seq(
+    O.create[p1.x.type](p1.x),
+    C.create[p3.x](new p3.x),
+
+    OO.create[p1.x.y.type](p1.x.y),
+    OC.create[p2.x.y](new p2.x.y),
+    CO.create[c1.y.type forSome { val c1: p3.x }]({ val c1 = new p3.x ; c1.y }),
+    CC.create[p4.x#y]({ val c1 = new p4.x ; new c1.y }),
+
+    OOO.create[p1.x.y.z.type](p1.x.y.z),
+    OOC.create[p1.x.y.z](new p1.x.y.z),
+    OCO.create[c2.z.type forSome { val c2: p2.x.y }]({ val c2 = new p2.x.y ; c2.z }),
+    OCC.create[p2.x.y#z]({ val c2 = new p2.x.y ; new c2.z }),
+    COO.create[c1.y.z.type forSome { val c1: p3.x }]({ val c1 = new p3.x ; c1.y.z }),
+    COC.create[c1.y.z forSome { val c1: p3.x }]({ val c1 = new p3.x ; val c2 = c1.y ; new c2.z }),
+    CCO.create[c2.z.type forSome { val c2: p4.x#y }]({ val c1 = new p4.x ; val c2 = new c1.y ; c2.z }),
+    CCC.create[p4.x#y#z]({ val c1 = new p4.x ; val c2 = new c1.y ; new c2.z })
+  )
+
+  def run(c: Class[_]) = strip(ClassToAPI.classCanonicalName(c))
+  def run2(c: Class[_]) = handleMalformed(strip(c.getCanonicalName))
+
+  def handleMalformed(s: => String) =
+    try s catch {
+      case e: InternalError if e.getMessage == "Malformed class name" => "Malformed class name"
+    }
+
+  def strip(s: String) = {
+    val prefix = "sbt.internal.inc.p"
+    if (s startsWith prefix) s stripPrefix prefix drop 2
+    else s
+  }
+
+  class Name(val value: String, malformedCN: Boolean = false) {
+    def create[T: ClassTag](ref: T): Data = {
+      val name = getClass.getSimpleName stripSuffix "$"
+      val canon = if (malformedCN) "Malformed class name" else value
+      Data(name, value, ref.getClass, classTag[T].runtimeClass, canon)
+    }
+  }
+
+  case class Data(nesting: String, full: String, refClass: Class[_], tagClass: Class[_], canon: String)
+}
+
+// All eight combinations of class/object nesting to three levels.
+// Because of scala/bug#2034 we can't put these all in one package (you get "name clash" errors)
+// So instead we'll split them in 4 nice and even packages
+package p1 { object x { object y { object z; class z } } }
```

Comment:
Not sure. We don't have neo-sbt-scalafmt ATM, and I don't have an up-to-date CLI launcher. Let me know what it should look like and I'll change it; meanwhile I've dropped the space tabbing I had... 😢